src/operator/aggregate.rs

```diff
@@ -112,63 +107,104 @@ where
+        Z::Val: Ord + Clone,
+        F: Fn(&Z::Key, &Z::Val) -> A + Clone + 'static,
+        O: Clone + ZSet<Key = (Z::Key, A), R = Z::R> + DeepSizeOf + 'static,
+        O::R: ZRingValue,
+        A: MulByRef<Z::R> + GroupValue + DeepSizeOf + Send,
+    {
+        self.weigh(f).aggregate_incremental_nested(|key, weights| {
+            debug_assert_eq!(weights.len(), 1);
+            (key.clone(), weights.pop().unwrap().1)
+        })
+    }
+
+    /// Convert indexed Z-set `Z` into a Z-set where the weight of each key
+    /// is computed as:
     ///
-    /// This method only works for linear aggregation functions `f`, i.e.,
-    /// functions that satisfy `f(a+b) = f(a) + f(b)`.  It will produce
-    /// incorrect results if `f` is not linear.
-    pub fn aggregate_linear_incremental_nested<K, VI, VO, W, F, O>(
-        &self,
-        f: F,
-    ) -> Stream<Circuit<P>, O>
+    /// ```text
+    ///    __
+    ///    ╲
+    ///    ╱ f(k,v) * w
+    ///    ‾‾
+    /// (k,v,w) ∈ Z
+    /// ```
+    ///
+    /// This is a linear operator.
+    pub fn weigh<F, T>(&self, f: F) -> Stream<Circuit<P>, OrdZSet<Z::Key, T>>
     where
-        K: KeyProperties,
-        VI: GroupValue,
-        SR: SharedRef + 'static,
-        <SR as SharedRef>::Target: ZSet<K, VI>,
-        <SR as SharedRef>::Target: NumEntries + SharedRef<Target = SR::Target>,
-        for<'a> &'a <SR as SharedRef>::Target: IntoIterator<Item = (&'a K, &'a VI)>,
-        F: Fn(&K, &VI) -> VO + 'static,
-        VO: NumEntries + GroupValue,
-        W: ZRingValue,
-        O: Clone + MapBuilder<(K, VO), W> + NumEntries + GroupValue,
+        Z: IndexedZSet,
+        Z::Key: Ord + Clone,
+        Z::Val: Ord + Clone,
+        F: Fn(&Z::Key, &Z::Val) -> T + 'static,
+        T: MulByRef<Z::R> + MonoidValue,
     {
-        self.integrate_nested()
-            .aggregate_linear_incremental(f)
-            .differentiate_nested()
+        self.weigh_generic::<_, OrdZSet<_, _>>(f)
+    }
+
+    /// Like [`Self::weigh`], but can return any batch type.
+    pub fn weigh_generic<F, O>(&self, f: F) -> Stream<Circuit<P>, O>
+    where
+        Z: IndexedZSet,
+        Z::Key: Ord + Clone,
+        Z::Val: Ord + Clone,
+        F: Fn(&Z::Key, &Z::Val) -> O::R + 'static,
+        O: Clone + Batch<Key = Z::Key, Val = (), Time = ()> + 'static,
+        O::R: MulByRef<Z::R>,
+    {
+        self.apply(move |batch| {
+            let mut delta = Vec::with_capacity(batch.key_count());
+            let mut cursor = batch.cursor();
+            while cursor.key_valid() {
+                let mut agg = HasZero::zero();
+                while cursor.val_valid() {
+                    agg += f(cursor.key(), cursor.val()).mul_by_ref(&cursor.weight());
+                    cursor.step_val();
+                }
+                delta.push((cursor.key().clone(), agg));
+                cursor.step_key();
+            }
+            O::from_keys((), delta)
```

Comment:
Can't this use a builder since the key is unchanged?

Reply:
Great idea!